### DISCUSSION TITLE
Best practices for nested settings route

### DISCUSSION THREAD

**Discussion by segersniels** [remix-run/remix#9725] [Q&A] [answered]
Imagine a SPA with a settings sheet similar to shadcdn's https://ui.shadcn.com/docs/components/sheet but as a nested route under `/settings`. With the way that Remix routing works the `/settings` route targets the root layout at `app/root.tsx` which removes all content from `app/routes/_index.tsx` since it's not rendered through an `<Outlet />` there.

Now my current solution (which you can see in action [here](https://chat.niels.foo) if you press the sliders icon in the sidebar) basically moved my entire page from `_index.tsx` under `app/root.tsx` and removed the `_index.tsx` route entirely. But this feels very counter intuitive that my entire app would be under the root route incl. the actions and loaders associated with the route. Something feels 💩 

My DX oriented mind would love a way to keep the root route super basic with just the sidebar and outlet and still have the `_index.tsx` containing the `/` route code. Is there a way I can tell a nested route to target a specific outlet? Meaning that I could tell `/settings` to render in an `<Outlet />` of `_index.tsx` instead of the `root`.

Another way I could think of was creating a layout route `_.tsx` with the entire `/` code and then have a `_.settings.tsx` with the nested route. But not sure if I'm a fan of that either. Cause what do you even do to handle the index route? Create a dummy `_._index.tsx` route that just returns `null`? 

What are the usual best practices for UI like this? Or am I approaching this wrong entirely and should I just not use nested routes for something like this.

**Comment by segersniels** [comment]
Ok thinking about it from a different angle about a project with more than just the root route, having a nested route for something like this wouldn't even work. I guess I have my answer there why something like this shouldn't be a nested route.

**Reply by kiliman** [reply]
Here's a sample app that may help.

1. Remix doesn't support child routes under the `index` route, so I named it _home_ and added a redirect from `/` to `/home`.

2. I'm using a React portal to render the panel. See _components/panel_ and _components/sidebar_. The _root_ route renders the `<Outlet>` and `<Sidebar>`

3. The _home.settings_ route is an actual Remix route, so it can have a loader and action. The _home_ route also has an `<Outlet>` to render the settings panel.

⚡️ StackBlitz https://stackblitz.com/edit/remix-run-remix-zqgjfg

**Reply by segersniels** [reply]
That's one way of doing it indeed which I also thought about, with a redirect on the root route. But then again when you start introducing other routes, like let's say `/about`. You would be navigating to `/home/settings` and rendering the `/home` content underneath. I guess that shows that using nested routes for this is not an ideal pattern and it's better to just have the sidebar render an overlayed absolute `div` over w/e content the page is currently on without navigating.

**Reply by kiliman** [reply]
Ok, I updated the example app.

I used the Remix `routes` config to manually add the _panel_ routes (in this case _settings_) to the top-level routes. So Remix now sees `/home/settings` and `/about/settings` and they both use the same route file.
```ts
    remix({
      ignoredRouteFiles: ['**/*'],
      routes: () => {
        const routes = flatRoutes('app');
        addPanel(routes, 'routes/home', 'settings');
        addPanel(routes, 'routes/about', 'settings');
        console.log(routes);
        return routes;
      },
    // snip

function addPanel(routes: any, parentId: string, panel: string) {
  const panelId = `${parentId}/${panel}`;
  routes[panelId] = {
    id: panelId,
    parentId,
    path: panel,
    file: `routes/panels/${panel}.tsx`,
  };
}
```
```ts
<Routes>
  <Route file="root.tsx">
    <Route index file="routes/_index.tsx" />
    <Route path="about" file="routes/about.tsx">
      <Route path="settings" file="routes/panels/settings.tsx" />
    </Route>
    <Route path="home" file="routes/home.tsx">
      <Route path="settings" file="routes/panels/settings.tsx" />
    </Route>
  </Route>
</Routes>
```

**Reply by segersniels** [reply]
That's pretty cool. Didn't know you could do that, TIL! Thanks! Is that a common design pattern to do it like that in Remix?

**Reply by kiliman** [reply]
I don't know about common, but that is why they opened up the `routes` config so you can support whatever functionality you need.

I updated the example to also let you use the `_index` route with child routes. 

```ts
 routes: () => {
        const routes = flatRoutes('app');
        // delete the `index` prop from the index route
        // so we can have nested routes for panels
        routes['routes/_index'].index = false;
        routes['routes/_index'].path = '/';

        addPanel(routes, 'routes/_index', 'settings');
        addPanel(routes, 'routes/about', 'settings');

        console.log(routes);
        return routes;
      },
```

**Reply by kiliman** [reply, marked as answer]
I also fixed the issue where the settings panel didn't render properly if you refreshed on the `/about/settings` route. It was a hydration issue that was fixed using the RC versions of `react` and `react-dom` (v19).
